Enable tests failing due to missing array ctor and lbound index checks.

lateralusX opened the pull request:
Fix https://github.com/mono/mono/pull/19494 adds missing array ctor needed by several tests currently failing as described by https://github.com/dotnet/runtime/issues/34068.

This PR fixes an additional problem hit by the IL array tests using an lbound != 0. Current Mono codegen for Set/Get/Address array methods didn't adjust index with regards to lbound causing failures in several of the array tests. NOTE, this scenario should only be hit when directly calling Set/Get/Address using IL on an array with an lbound != 0. It shouldn't affect regular szarray's at all since that code paths should never be used and the codegen currently looks up the lbounds unconditionally in the array, so if that code path ever gets hit by a szarray, it will imdediate crash, so shouldn't be causing any performance impact on szarrays. There is also an additional check on the array type as well as an additional precaution, so only arrays of type MONO_TYPE_ARRAY, hitting codegen for Set/Get/Address array method calls will adjust index based on lbound.

This PR fix the lbound index codegen for that specific code path and re-eanbles the tests currently listed in https://github.com/dotnet/runtime/issues/34068.

danmosemsft commented:
Poor bot, I guess it saw ".targets"

lateralusX commented:
Depending on https://github.com/dotnet/runtime/pull/34054 to be merged before CoreCLR test suites will run using Mono on CI.

lateralusX commented:
@naricc CoreCLR tests are now running on Mono runtime, correct? If so I guess it's time to rebase this PR and validate the results when reenable the array ctor tests.

naricc commented:
@lateralusX Yes, the runtime tests (I'm trying to stop calling them the CoreCLR tests, since they now are testing multiple runtimes and it gets confusing :) ) are now in CI. You will see them as "Mono Pri0 Runtime Tests Run OSX x64 release".  I am working on standing up CI lanes for other architectures/modes (like interpreter).

lambdageek commented:
@naricc could you resolve the conflicts in this PR

naricc commented:
@lambdageek Ok; working on it.

stephentoub commented:
There are conflicts. Mind rebasing? Thanks.

lateralusX commented:
Investigate this a little further, we now get passed the error fixed by https://github.com/mono/mono/pull/19494, but several of these tests also use constructions to test lbound != 0 and then access the array using indexes inline with the new low and upper bounds. On Mono this will fail with 1 dim arrays since we don't adjust index based on lbound != 0. This needs to be fixed before enable the tests or we will still get failures.

stephentoub commented:
Thanks.  Is there an open issue tracking that?

stephentoub commented:
(And should this be closed in the interim?)

lateralusX commented:
I believe the only one we currently have is the one referenced in this PR, https://github.com/dotnet/runtime/issues/34068, but that highlights another problem that is solved (missing constructor). I look into a potential fix for the other things, but depending on the outcome of that fix and if we could minimize perf impact, I could add that fix into this PR and re-enable tests. If we come to the conclusion that the fix adds to much perf overhead, I believe we also decided that we won't fix the failing tests, so then we can go ahead and close both this PR and the original issue.

lateralusX commented:
I will rebase this PR, add additional fix into this PR and re-enable all the tests tomorrow morning.

lateralusX commented:
@naricc I take over this PR again since it needs some additional fixes before all tests pass (not enough to just enable them).

lateralusX committed:
Fix index adjustments for Get/Set/Address called on 1 dim arrays.

Enable tests failing due to missing array ctor and lbound index checks.

Fix interpreter Get/Set/Address using lbound.

Add missing OP_FCONV_TO_I on Arm64, causing test errors.

Fix interpreter arrray index and lower bound when using negative values.

Add branch checking array->bounds for NULL.

vargaz commented on src/mono/mono/mini/method-to-ir.c:
```diff
@@ -4078,11 +4078,11 @@ mini_emit_sext_index_reg (MonoCompile *cfg, MonoInst *index)
 MonoInst*
-mini_emit_ldelema_1_ins (MonoCompile *cfg, MonoClass *klass, MonoInst *arr, MonoInst *index, gboolean bcheck)
+mini_emit_ldelema_1_ins (MonoCompile *cfg, MonoClass *klass, MonoInst *arr, MonoInst *index, gboolean bcheck, gboolean adjust_index)
 {
```
'bounded' might be a better param name.

lateralusX replied:
Fixed.

lateralusX commented:
Test pass on Windows x64, so expected at least OSX x64 to get same results, but test runs looks like the additional codegen is not hit, since failures looks close to what happens when the lbound is not adjusted. Need to investigate the tests on none Windows platforms.

lateralusX commented:
Failures was not in JIT, but interpreter config, updated with similar adjustments done to JIT in interpreter.

lateralusX commented:
There is also a missing OP code on arm64 causing issue to convert float -> int, adding simliar conversion for OP_FCONV_TO_I as done on amd64, arm and x86.

lateralusX commented:
Down to 3 failures, only interpreter lane on _il_relgcarr, _il_relint32_m1, _il_dbgint32_m1, so these three cases must hit some other code path in interpreter compared to rest, will investigate.

lateralusX commented:
Down to one failing tests in mono/mono, apparently we take one internal shortcut when creating a MONO_TYPE_ARRAY with lower bound of 0 and don't create the runtime lbound struct, so we need to add condition in codegen for that case.
